### PR TITLE
[MIG][12.0] fieldservice_helpdesk_stock

### PR DESCRIPTION
**Pull request by murtuzasaleh**
Depends on

- [x] #82
- [x] #83

### CODE REVIEW

**Review by osi-scampbell — DISMISSED** [review]

**Review by max3903 — DISMISSED** [review]

**Comment by max3903** [comment]
@murtuzasaleh Please rebase and add this module to the exclude list.

**Comment by murtuzasaleh** [comment]
@max3903 It is in WIP, I will work on it after completed fieldservice_stock module migration part.

**Comment by max3903** [comment]
There is 2 flows:

1. The helpdesk user creates a stock request from the helpdesk ticket.
2. The FSM user creates a stock request from the fsm order.

If the fsm order is linked to a helpdesk ticket, then the stock request should be linked to the fsm order AND the related ticket, so the helpdesk user can see all the stock requests related to the ticket even if they have been created from one of the related fsm orders.

The first user (Helpdesk or FSM) who submits the stock request wins.

**Comment by murtuzasaleh** [comment]
@max3903 In FSM Order << Can I add any direction for helpdesk line route?
In the direction, There are two value (outbound and inbound) for FSM order

**Comment by max3903** [comment]
@murtuzasaleh Sure.

**Comment by murtuzasaleh** [comment]
> @murtuzasaleh Sure.

@max3903 `Helpdesk Line` Name is ok?

**Comment by max3903** [comment]
@murtuzasaleh I am not sure to follow you here.

In helpdesk_fieldservice, we have a ticket_id field on fsm.order.
In helpdesk_stock, we have a ticket_id field on stock.request and a request_ids on helpdesk.ticket
In fieldservice_stock, we have a fsm_order_id field on stock.request and a request_ids on fsm.order.
In fieldservice_helpdesk_stock, when a stock request is created from the fsm.order, we should set stock_request.ticket_id = fsm_order.ticket_id. --> There is no UI changes, so where do you want to set this name???

**Comment by murtuzasaleh** [comment]
> fsm_order_id

Sorry, my mistake.

**Comment by murtuzasaleh** [comment]
> ticket_id

@max3903 when I have assigned ticket_id in fsm.order based on refactored code I found `Stock Request name must be unique ` .
Can you please guide me on this.

`class FSMOrder(models.Model):
    _inherit = 'fsm.order'

    def action_inventory_confirm(self):
        res = super(FSMOrder, self).action_inventory_confirm()
        for order in self:
            order.procurement_group_id.helpdesk_ticket_id = order.ticket_id
        return res

    @api.multi
    def write(self, vals):
        res = super(FSMOrder, self).write(vals)
        if 'ticket_id' in vals:
            for line in self.stock_request_ids:
                if line.state in ('draft', 'cancelled'):
                    if line.ticket_id:
                        line.ticket_id.ticket_id = \
                            vals.get('ticket_id')
                    else:
                        line.ticket_id = \
                            self.env['stock.request'].create({
                                'product_id': line.product_id.id,
                                'name': line.name,
                                'product_uom_qty': line.product_uom_qty,
                                'product_uom_id': line.product_uom_id.id,
                                'ticket_id': vals.get('ticket_id'),
                                'location_id': line.location_id.id,
                                'warehouse_id': line.warehouse_id.id,
                                # 'fsm_order_line_id': line.id,
                                'state': line.state,
                            })
        return res`

Is it right?

**Comment by max3903** [comment]
@murtuzasaleh You are creating a new stock_request and line.ticket_id expect the id of a ticket, not the one of a stock.request.

If you change the ticket of an order, you should update (not create) the related stock request with the new ticket:

# Before
Order 1
* Ticket 2
* Stock Requests
  * Request 1: Ticket 2
  * Request 2: Ticket 2

# After
Order 1
* Ticket 3
* Stock Requests
  * Request 1: Ticket 3
  * Request 2: Ticket 3

**Comment by murtuzasaleh** [comment]
@max3903 can you please review this code,
I think no need for fieldservice_helpdesk_stock/models/helpdesk_ticket.py file, correct me If I am wrong.

File: fieldservice_helpdesk_stock/models/fsm_order.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    @api.multi
+    def write(self, vals):
+        res = super(FSMOrder, self).write(vals)
+        for rec in self:
+            if 'ticket_id' in vals:
+                for line in self.stock_request_ids:
+                    if line.state in ('draft', 'cancelled'):
+                        if line.ticket_id:
+                            line.ticket_id.ticket_id = \
```
Comment (max3903) [inline fieldservice_helpdesk_stock/models/fsm_order.py]:
```suggestion
                            line.ticket_id = \
```

File: fieldservice_helpdesk_stock/models/fsm_order.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+                                vals.get('ticket_id')
+                        else:
+                            rec.ticket_id = vals.get('ticket_id')
```
Comment (max3903) [inline fieldservice_helpdesk_stock/models/fsm_order.py]:
```suggestion
                    if line.state in ('draft', 'cancelled'):
                        line.ticket_id = vals.get('ticket_id')
```

**Comment by max3903** [comment]
@murtuzasaleh Correct.

**Comment by murtuzasaleh** [comment]
@osi-scampbell Can you please review this PR

**Comment by murtuzasaleh** [comment]
@max3903 Can you please review this PR.

**Review by max3903 — APPROVED** [review]

**Review by osi-scampbell — APPROVED** [review]

**Pull request merged**